Codec/Archive/Tar/Types.hs

```diff
@@ -358,83 +377,111 @@ instance Show TarPath where
 --
 -- The difference compared to 'fromTarPath' is that it always returns a Unix
 -- style path irrespective of the current operating system.
 --
 -- This is useful to check how a 'TarPath' would be interpreted on a specific
 -- operating system, eg to perform portability checks.
 --
-fromTarPathToPosixPath :: TarPath -> FilePath
-fromTarPathToPosixPath = BS.Char8.unpack . fromTarPathInternal FilePath.Posix.pathSeparator
+fromTarPathToPosixPath :: TarPath -> PosixPath
+fromTarPathToPosixPath (TarPath name prefix)
+  | PS.null prefix = name
+  | PS.null name = prefix
+  | otherwise = prefix <> PS.cons PFP.pathSeparator name
 
--- | Convert a 'TarPath' to a Windows 'FilePath'.
+-- | Convert a 'TarPath' to a Windows 'OsPath'.
 --
 -- The only difference compared to 'fromTarPath' is that it always returns a
 -- Windows style path irrespective of the current operating system.
 --
 -- This is useful to check how a 'TarPath' would be interpreted on a specific
 -- operating system, eg to perform portability checks.
 --
-fromTarPathToWindowsPath :: TarPath -> FilePath
-fromTarPathToWindowsPath = BS.Char8.unpack . fromTarPathInternal FilePath.Windows.pathSeparator
-
-fromTarPathInternal :: Char -> TarPath -> BS.ByteString
-fromTarPathInternal sep (TarPath name prefix)
-  | BS.null prefix = adjustSeps name
-  | BS.null name = adjustSeps prefix
-  | sep == FilePath.Posix.pathSeparator = prefix <> BS.Char8.cons sep name
-  | otherwise = adjustSeps prefix <> BS.Char8.cons sep (adjustSeps name)
-  where
-    adjustSeps = BS.Char8.map $ \c -> if c == FilePath.Posix.pathSeparator then sep else c
-{-# INLINE fromTarPathInternal #-}
-
--- | Convert a native 'FilePath' to a 'TarPath'.
---
--- The conversion may fail if the 'FilePath' is empty or too long.
+fromTarPathToWindowsPath :: MonadThrow m => TarPath -> m WindowsPath
+fromTarPathToWindowsPath tarPath = do
+  let posix = fromTarPathToPosixPath tarPath
+  toWindowsPath posix
+
+-- | We assume UTF-8 on posix and UTF-16 on windows.
+toWindowsPath :: MonadThrow m => PosixPath -> m WindowsPath
+toWindowsPath posix = do
+  str <- PFP.decodeUtf posix
+  win <- WFP.encodeUtf str
+  pure $ WS.map (\c -> if WFP.isPathSeparator c then WFP.pathSeparator else c) win
+
+-- | We assume UTF-8 on posix and UTF-16 on windows.
+toPosixPath :: MonadThrow m => WindowsPath -> m PosixPath
+toPosixPath win = do
+  str   <- WFP.decodeUtf win
+  posix <- PFP.encodeUtf str
+  pure $ PS.map (\c -> if PFP.isPathSeparator c then PFP.pathSeparator else c) posix
+
+-- | We assume UTF-8 on posix and UTF-16 on windows.
+toPosixPath' :: MonadThrow m => OsPath -> m PosixPath
+#if defined(mingw32_HOST_OS)
+toPosixPath' (OsString ws) = toPosixPath ws
+#else
+toPosixPath' (OsString ps) = pure ps
+#endif
+
+-- | We assume UTF-8 on posix and UTF-16 on windows.
+fromPosixPath :: MonadThrow m => PosixPath -> m OsPath
+#if defined(mingw32_HOST_OS)
+fromPosixPath ps = OsPath <$> toWindowsPath ps
+#else
+fromPosixPath ps = pure $ OsString ps
+#endif
```

Comment:
These are the main conversion functions. As we can see... we leave posix filepaths untouched, but assume UTF-8 encoding when converting from posix filepaths (e.g. those coming from the actual tar archive) to windows, where we assume UTF-16.

Reply:
The tar spec obviously demands `PosixPath` where we don't assume an encoding. So all filepaths within the tar archive are posix.

Reply:
Yep, makes sense.